torch_stoi/stoi.py

```diff
@@ -192,55 +238,97 @@ def detect_silent_frames(x, dyn_range, framelen, hop):
         Returns:
             torch.BoolTensor, framewise mask.
         """
+        window = torch.from_numpy(np.hanning(framelen + 2)[1:-1])
+
         x_frames = unfold(x[:, None, None, :], kernel_size=(1, framelen),
-                          stride=(1, hop))[..., :-1]
+                          stride=(1, hop))
+        y_frames = unfold(y[:, None, None, :], kernel_size=(1, framelen),
+                          stride=(1, hop))
+        x_frames *= window[:, None]
+        y_frames *= window[:, None]
+
         # Compute energies in dB
         x_energies = 20 * torch.log10(torch.norm(x_frames, dim=1,
                                                  keepdim=True) + EPS)
         # Find boolean mask of energies lower than dynamic_range dB
         # with respect to maximum clean speech energy frame
-        mask = (torch.max(x_energies, dim=2, keepdim=True)[0] - dyn_range -
-                x_energies) < 0
-        return mask
+        mask = (x_energies.amax(2, keepdim=True) - dyn_range - x_energies) < 0
+        mask = mask.squeeze(1)
+
+        # Remove silent frames and pad with zeroes
+        x_frames = x_frames.permute(0, 2, 1)
+        y_frames = y_frames.permute(0, 2, 1)
+        x_frames = _mask_audio(x_frames, mask)
+        y_frames = _mask_audio(y_frames, mask)
+
+        x_sil = _overlap_and_add(x_frames, hop)
+        y_sil = _overlap_and_add(y_frames, hop)
+        x_frames = x_frames.permute(0, 2, 1)
+        y_frames = y_frames.permute(0, 2, 1)
+
+        return x_sil, y_sil, mask
 
     @staticmethod
     def stft(x, win, fft_size, overlap=4):
+        """We can't use torch.stft:
+        - It's buggy with center=False as it discards the last frame
+        - It pads the frame left and right before taking the fft instead
+        of padding right
+        - Probably other reasons; torch.stft is very broken and too abstract
```

Comment:
Can remove this sentence.

Reply:
The last one or all three lines? I think the first two are valuable. I can remove the last one.